refl1d/reflectivity.py

```diff
@@ -285,33 +298,340 @@ def calculate_u1_u3_py(H, rhoM, thetaM, Aguide):
 
     u1 = u1_num/u1_den
     u3 = u3_num/u3_den
-    #print "u1", u1
-    #print "u3", u3
+    # print "u1", u1
+    # print "u3", u3
     return sld_b, u1, u3
 
+
 try:
-    #raise ImportError() # uncomment to force numba off
-    from numba import njit
+    # raise ImportError() # uncomment to force numba off
+    from numba import njit, prange
     USE_NUMBA = True
 except ImportError:
     USE_NUMBA = False
     # if no numba then njit does nothing
+
     def njit(*args, **kw):
         # Check for bare @njit, in which case we just return the function.
         if len(args) == 1 and callable(args[0]) and not kw:
             return args[0]
         # Otherwise we have @njit(...), so return the identity decorator.
         return lambda fn: fn
 
-@njit('(f8[:], f8[:], f8[:], f8[:], f8[:])')
+MINIMAL_RHO_M = 1e-2  # in units of 1e-6/A^2
+EPS = np.finfo(float).eps
+B2SLD = 2.31604654  # Scattering factor for B field 1e-6
+
+from numba.pycc import CC
+cc = CC('magnetic_amplitude')
```

Comment:
Any reason to do AOT compile rather than simply shipping numba and letting it use whatever optimization is available for the platform it is running on? Sounds like less work to me. Startup times will be slightly slower, but not an issue when fitting since the compile only needs to happen once. Or maybe once per process: I don't know how numba interacts with multiprocessing.

Note that numba is able to parallelize across cores and sockets on the same machine, so it may be faster to run bumps single threaded with the nllf running in parallel rather than running the optimizer in parallel with the nllf single threaded. Or maybe fewer threads than cores to allow some overlap during the non-parallel bits while reducing contention during the parallel bits.

Reply:
The startup times were pretty long when using njit without AOT... 10+ seconds on my machine.  There are probably things that can be done to speed this up.  I found that for whatever reason the njit code was slightly slower than the AOT code, though that's not supposed to be true.

Reply:
Caching should take care of that. You are doing this by hand anyway, so let numba take care of it. It seems to work for me. It took 36s to run check_examples.py with njit alone and 38s to run with precompiled, in both cases having updated the cache prior to timing.

Reply:
I found that it doesn't cache between starts of refl1d - it recompiles every time the program is started, which is potentially very annoying.

Reply:
That hasn't been my experience (mac conda, python=3.8, numba=0.53.1). If you touch the file it'll recompile, but otherwise it should use the cache. Any chance you are touching the .py file between runs? [obviously not, but I have to ask...]

There are some issues on [this](https://github.com/numba/numba/issues/2439) from [two years ago](https://stackoverflow.com/questions/56936573/caching-jit-compiled-functions-in-numba). Perhaps they weren't resolved?

Are there any warnings showing up if you run from the command line?

The [docs](https://numba.pydata.org/numba-doc/latest/developer/caching.html) tell you how to set the cache directory. Can you check that the cache is being created?

Reply:
You are correct - the cache seems to be working.  I think I was thrown off because I was editing some other parts of the source file (not in the numba functions) and I didn't think that would affect the cache, but it does.